src/shogun/neuralnets/NeuralRecurrentLayer.h

```diff
@@ -0,0 +1,220 @@
+ */
+
+#ifndef __NEURALRECURRENTLAYER_H__
+#define __NEURALRECURRENTLAYER_H__
+
+#include <shogun/lib/common.h>
+#include <shogun/neuralnets/NeuralLinearLayer.h>
+
+namespace shogun
+{
+/** @brief Recurrent Neural layer with linear neurons, with an identity activation
+ * function. can be used as a hidden layer or an output layer
+ *
+ * Each neuron in the layer is connected to all the neurons in all the input
+ * and hidden layers that connect into this layer.
+ *
+ * The hidden state at time step t is computed according to the previous hidden state
+ * and the input at the current state as:
+ *
+ * \f$ h_t = f(W x_t + U s_t_-_1 + b_h) \f$ where \f$ b_h \f$ is the bias vector of
+ * the hidden layer, \f$ W \f$ is the weights matrix between this layer and the input layer,
+ * and \f$ U \f$ is the weights matrix between this hidden layer and itself
+ *
+ * The output layer is computed as:
+ *
+ * \f$ y_t = f(V h_t + b_y) \f$ where \f$ b_y \f$ is the bias vector of the output layer,
+ * \f$ h_t \f$ is the value of the hidden states at the current time step and \f$ V \f$ is
+ * the weights matrix between this hidden layer and the output layer
+ *
+ */
+class CNeuralRecurrentLayer : public CNeuralLinearLayer
+{
+public:
+	/** default constructor */
+	CNeuralRecurrentLayer();
+
+	/** Constuctor
+	 *
+	 * @param num_neurons Number of neurons in this layer
+	 * @param time_series_length length of time series for this layer
+	 * @param output_dim outpit size of this layer
+	 */
+	CNeuralRecurrentLayer(int32_t num_neurons, int32_t time_series_length, int32_t output_dim);
+
+	virtual ~CNeuralRecurrentLayer() {}
+
+	/** Initializes the layer, computes the number of parameters needed for
+	 * the layer
+	 *
+	 * @param layers Array of layers that form the network that this layer is
+	 * being used with
+	 *
+	 * @param input_indices  Indices of the layers that are connected to this
+	 * layer as input
+	 */
+	virtual void initialize_neural_layer(CDynamicObjectArray* layers,
+			SGVector<int32_t> input_indices);
+
+	/** Sets the batch_size and allocates memory for m_activations and
+	 * m_input_gradients accordingly. Must be called before forward or backward
+	 * propagation is performed
+	 *
+	 * @param batch_size number of training/test cases the network is
+	 * currently working with
+	 */
+	virtual void set_batch_size(int32_t batch_size);
+
+	/** Initializes the layer's parameters. The layer should fill the given
+	 * arrays with the initial value for its parameters
+	 *
+	 * @param parameters Vector of size get_num_parameters()
+	 *
+	 * @param parameter_regularizable Vector of size get_num_parameters().
+	 * This controls which of the layer's parameter are
+	 * subject to regularization, i.e to turn off regularization for parameter
+	 * i, set parameter_regularizable[i] = false. This is usally used to turn
+	 * off regularization for bias parameters.
+	 *
+	 * @param sigma standard deviation of the gaussian used to random the
```

Comment:
Capital S

Reply:
A few comments

* "random" is not a verb. Should be be explicit about what actually is done
* What if I want non-Gaussian noise to be added? Shouldnt this be more modular?